xrootd/cmd/xrd-srv/handler.go

```diff
@@ -48,3 +60,153 @@ func (h *handler) Dirlist(sessionID [16]byte, request *dirlist.Request) (xrdprot
 
 	return resp, xrdproto.Ok
 }
+
+// Open implements server.Handler.Open.
+func (h *handler) Open(sessionID [16]byte, request *open.Request) (xrdproto.Marshaler, xrdproto.ResponseStatus) {
+	var flag int
+	if request.Options&xrdfs.OpenOptionsOpenRead != 0 {
+		flag |= os.O_RDONLY
+	}
+	if request.Options&xrdfs.OpenOptionsOpenUpdate != 0 {
+		flag |= os.O_RDWR
+	}
+	if request.Options&xrdfs.OpenOptionsOpenAppend != 0 {
+		flag |= os.O_APPEND
+	}
+	if request.Options&xrdfs.OpenOptionsNew != 0 || request.Options&xrdfs.OpenOptionsDelete != 0 {
+		flag |= os.O_CREATE
+		if request.Options&xrdfs.OpenOptionsDelete == 0 {
+			flag |= os.O_EXCL
+		} else {
+			flag |= os.O_TRUNC
+		}
+	}
+
+	filePath := path.Join(h.basePath, request.Path)
+	if request.Options&xrdfs.OpenOptionsMkPath != 0 {
+		if err := os.MkdirAll(path.Dir(filePath), os.FileMode(request.Mode)); err != nil {
+			return xrdproto.ServerError{
+				Code:    xrdproto.IOError,
+				Message: fmt.Sprintf("An IO error occurred: %v", err),
+			}, xrdproto.Error
+		}
+	}
+
+	file, err := os.OpenFile(filePath, flag, os.FileMode(request.Mode))
+	if err != nil {
+		return xrdproto.ServerError{
+			Code:    xrdproto.IOError,
+			Message: fmt.Sprintf("An IO error occurred: %v", err),
+		}, xrdproto.Error
+	}
+
+	sess, _ := h.sessions.LoadOrStore(sessionID, &session{handles: make(map[xrdfs.FileHandle]*os.File)})
+	s := sess.(*session)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	var handle xrdfs.FileHandle
+
+	for i := 0; i < 255; i++ {
```

Comment:
this double-loop needs some explanation:
- why are we doing this
- what does it represent
- what is allowed to only use the first 2 bytes of the file handle as a tag
- etc...

Reply:
good catch! I actually messed up with it since `FileHandle` is `[4]byte` and not `[2]byte`, my bad.

We need to generate unique handle somehow here and now I wonder if it will be more effective to `rand.Read` 4 bytes few times until we find a free handle...

What do you think?

Reply:
we could either do the same trick than in `mux` (and keep a per-handler `uint32` counter) or just `rand.Read([4]byte[:])`.
not sure what the predictability would mean in terms of security. (I have next to zero knowledge in that domain.)

I guess `crypto/rand.Read` is fine for now (but please document this in the code so "future maintainer" will know what she needs to pay attention to.)